pytorchfi/errormodels.py

```diff
@@ -0,0 +1,358 @@
+        (conv, C, H, W) = random_neuron_location(pfi_model)
+    if randVal is False:
+        err_val = random_value(min_val=min_val, max_val=max_val)
+
+    for i in range(pfi_model.get_total_batches()):
+        if randLoc is True:
+            (conv, C, H, W) = random_neuron_location(pfi_model)
+        if randVal is True:
+            err_val = random_value(min_val=min_val, max_val=max_val)
+
+        batch.append(i)
+        conv_num.append(conv)
+        c_rand.append(C)
+        h_rand.append(H)
+        w_rand.append(W)
+        value.append(err_val)
+
+    return pfi_model.declare_neuron_fi(
+        batch=batch, conv_num=conv_num, c=c_rand, h=h_rand, w=w_rand, value=value
+    )
+
+
+# one random neuron error per layer in single batch element
+def random_inj_per_layer(pfi_model, min_val=-1, max_val=1):
+    batch, conv_num, c_rand, h_rand, w_rand, value = ([] for i in range(6))
+
+    b = random_batch_element(pfi_model)
+    for i in range(pfi_model.get_total_conv()):
+        (C, H, W) = random_neuron_location_conv(pfi_model, i)
+        batch.append(b)
+        conv_num.append(i)
+        c_rand.append(C)
+        h_rand.append(H)
+        w_rand.append(W)
+        value.append(random_value(min_val=min_val, max_val=max_val))
+
+    return pfi_model.declare_neuron_fi(
+        batch=batch, conv_num=conv_num, c=c_rand, h=h_rand, w=w_rand, value=value
+    )
+
+
+# one random neuron error per layer in each batch element
+def random_inj_per_layer_batched(
+    pfi_model, min_val=-1, max_val=1, randLoc=True, randVal=True
+):
+    batch, conv_num, c_rand, h_rand, w_rand, value = ([] for i in range(6))
+
+    for i in range(pfi_model.get_total_conv()):
+        if randLoc is False:
+            (C, H, W) = random_neuron_location_conv(pfi_model, i)
+        if randVal is False:
+            err_val = random_value(min_val=min_val, max_val=max_val)
+
+        for b in range(pfi_model.get_total_batches()):
+            if randLoc is True:
+                (C, H, W) = random_neuron_location_conv(pfi_model, i)
+            if randVal is True:
+                err_val = random_value(min_val=min_val, max_val=max_val)
+
+            batch.append(b)
+            conv_num.append(i)
+            c_rand.append(C)
+            h_rand.append(H)
+            w_rand.append(W)
+            value.append(err_val)
+
+    return pfi_model.declare_neuron_fi(
+        batch=batch, conv_num=conv_num, c=c_rand, h=h_rand, w=w_rand, value=value
+    )
+
+
+class single_bit_flip_func(core.fault_injection):
+    def __init__(self, model, h, w, batch_size, **kwargs):
+        super().__init__(model, h, w, batch_size, **kwargs)
+        self.bits = kwargs.get("bits", 8)
+        self.LayerRanges = []
+
+    def set_conv_max(self, data):
+        self.LayerRanges = data
+
+    def reset_conv_max(self, data):
+        self.LayerRanges = []
+
+    def get_conv_max(self, layer):
+        return self.LayerRanges[layer]
+
+    def _twos_complement(self, val, nbits):
```

Comment:
Duplicate function? lets clean it up in this PR to avoid extra diffs and changes in history :)

Reply:
Slightly different. Fixed by renaming and calling subroutine.